transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -192,86 +386,355 @@ public ChannelFuture deregister() {
+        }
+        write(msg, false, promise);
+
         return promise;
     }
 
+    private void invokeWrite(Object msg, ChannelPromise promise) {
+        try {
+            ((ChannelOutboundHandler) handler()).write(this, msg, promise);
+        } catch (Throwable t) {
+            notifyOutboundHandlerException(t, promise);
+        }
+    }
+
     @Override
     public ChannelHandlerContext flush() {
-        AbstractChannelHandlerContext next = findContextOutbound();
-        next.invokeFlush();
+        final AbstractChannelHandlerContext next = findContextOutbound();
+        EventExecutor executor = next.executor();
+        if (executor.inEventLoop()) {
+            next.invokeFlush();
+        } else {
+            Runnable task = next.invokeFlushTask;
+            if (task == null) {
+                next.invokeFlushTask = task = new Runnable() {
+                    @Override
+                    public void run() {
+                        next.invokeFlush();
+                    }
+                };
+            }
+            safeExecute(executor, task, channel().voidPromise(), null);
+        }
+
         return this;
     }
 
+    private void invokeFlush() {
+        try {
+            ((ChannelOutboundHandler) handler()).flush(this);
+        } catch (Throwable t) {
+            notifyHandlerException(t);
+        }
+    }
+
     @Override
     public ChannelFuture writeAndFlush(Object msg, ChannelPromise promise) {
-        AbstractChannelHandlerContext next = findContextOutbound();
-        next.invokeWriteAndFlush(pipeline.touch(msg, next), promise);
+        if (msg == null) {
+            throw new NullPointerException("msg");
+        }
+
+        if (!validatePromise(promise, true)) {
+            ReferenceCountUtil.release(msg);
+            // cancelled
+            return promise;
+        }
+
+        write(msg, true, promise);
+
         return promise;
     }
 
+    private void write(Object msg, boolean flush, ChannelPromise promise) {
+        AbstractChannelHandlerContext next = findContextOutbound();
+        EventExecutor executor = next.executor();
+        if (executor.inEventLoop()) {
+            next.invokeWrite(msg, promise);
+            if (flush) {
+                next.invokeFlush();
+            }
+        } else {
+            AbstractWriteTask task;
+            if (flush) {
+                task = WriteAndFlushTask.newInstance(next, msg, promise);
+            }  else {
+                task = WriteTask.newInstance(next, msg, promise);
+            }
+            safeExecute(executor, task, promise, msg);
+        }
+    }
+
     @Override
     public ChannelFuture writeAndFlush(Object msg) {
         return writeAndFlush(msg, newPromise());
     }
 
+    private static void notifyOutboundHandlerException(Throwable cause, ChannelPromise promise) {
+        if (!promise.tryFailure(cause) && !(promise instanceof VoidChannelPromise)) {
+            if (logger.isWarnEnabled()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1066)